File: java-solid-principle-sample-code/src/com/principle/solid/liskov_substitution/NoShapeV2.java
package com.principle.solid.liskov_substitution;

/**
	Liskov Substitution Principle Achieved - using `Abstract` Class for Substitution
*/
//public class NoShape implements Shape{
public class NoShapeV2 extends AbstractShape{

	/** 
		Liskov Substitution Principle violation - It is Substitution able, Even if we don't override.
        We have to either provide Interface `Default` implementation /or/ `Abstract` Class for Substitution
	 */
//    @Override
//    public double area() {
//        throw new IllegalStateException("Cannot calculate area");
//    }
}

File: python-solid-principle-sample-code/com/principle/solid/liskov_substitution/NoShapeV2.py
from Shape import Shape
from AbstractShape import AbstractShape
from IllegalStateException import IllegalStateException

"""
    Liskov Substitution Principle Achieved - using `Abstract` Class for Substitution
"""


# class NoShapeV2(Shape):
class NoShapeV2(AbstractShape):

    def __init__(self):
        pass

    """
        Liskov Substitution Principle violation - It is Substitution able, Even if we don't override.
        We have to either provide Interface `Default` implementation /or/ `Abstract` Class for Substitution
    """
    # def area(self):
    # raise IllegalStateException("Cannot calculate area")
